parsons/databases/bigquery/bigquery.py

```diff
@@ -0,0 +1,1123 @@
+                bigquery.SourceFormat.CSV
+                if kwargs["data_type"] == "csv"
+                else bigquery.SourceFormat.NEWLINE_DELIMITED_JSON
+            )
+
+        if not job_config.field_delimiter:
+            if kwargs["data_type"] == "csv":
+                job_config.field_delimiter = kwargs["csv_delimiter"]
+            if kwargs["nullas"]:
+                job_config.null_marker = kwargs["nullas"]
+
+        if not job_config.write_disposition:
+            if kwargs["if_exists"] == "append":
+                job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
+            elif kwargs["if_exists"] == "truncate":
+                job_config.write_disposition = bigquery.WriteDisposition.WRITE_TRUNCATE
+            elif kwargs["table_exists"] and kwargs["if_exists"] == "fail":
+                raise Exception("Table already exists.")
+            elif kwargs["if_exists"] == "drop" and kwargs["table_exists"]:
+                self.delete_table(kwargs["table_name"])
+                job_config.write_disposition = bigquery.WriteDisposition.WRITE_EMPTY
+            else:
+                job_config.write_disposition = bigquery.WriteDisposition.WRITE_EMPTY
+
+        if not job_config.allow_quoted_newlines:
+            job_config.allow_quoted_newlines = kwargs["allow_quoted_newlines"]
+
+        if kwargs["data_type"] == "csv" and kwargs["allow_jagged_rows"]:
+            job_config.allow_jagged_rows = kwargs["allow_jagged_rows"]
+        else:
+            job_config.allow_jagged_rows = True
+
+        if not job_config.quote_character and kwargs["quote"]:
+            job_config.quote_character = kwargs["quote"]
+
+        return job_config
+
+    def _fetch_query_results(self, cursor) -> Table:
+        # We will use a temp file to cache the results so that they are not all living
+        # in memory. We'll use pickle to serialize the results to file in order to maintain
+        # the proper data types (e.g. integer).
+        temp_filename = create_temp_file()
+
+        wrote_header = False
+        with open(temp_filename, "wb") as temp_file:
+            # Track whether we got data, since if we don't get any results we need to return None
+            got_results = False
+            while True:
+                batch = cursor.fetchmany(QUERY_BATCH_SIZE)
+                if len(batch) == 0:
+                    break
+
+                got_results = True
+
+                for row in batch:
+                    # Make sure we write out the header once and only once
+                    if not wrote_header:
+                        wrote_header = True
+                        header = list(row.keys())
+                        pickle.dump(header, temp_file)
+
+                    row_data = list(row.values())
+                    pickle.dump(row_data, temp_file)
+
+        if not got_results:
+            return None
+
+        ptable = petl.frompickle(temp_filename)
```

Comment:
What is the point of doing the pickling 100,000 at a time if we're just going to load them all back into memory at the end anyway?

Reply:
This was inherited from the previous version of BigQuery. Honestly didn't interrogate it much for improvements and I'm inclined to leave it for now